stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift

```diff
@@ -15,10 +15,133 @@ import SwiftPrivate
 import Darwin
 #elseif os(Linux) || os(FreeBSD) || os(PS4) || os(Android) || os(Cygwin) || os(Haiku)
 import Glibc
+#elseif os(Windows)
+import MSVCRT
+import WinSDK
 #endif
 
-
+internal func _signalToString(_ signal: Int) -> String {
+  switch CInt(signal) {
+  case SIGILL:  return "SIGILL"
+  case SIGABRT: return "SIGABRT"
+  case SIGFPE:  return "SIGFPE"
+  case SIGSEGV: return "SIGSEGV"
 #if !os(Windows)
+  case SIGTRAP: return "SIGTRAP"
+  case SIGBUS:  return "SIGBUS"
+  case SIGSYS:  return "SIGSYS"
+#endif
+  default:      return "SIG???? (\(signal))"
+  }
+}
+
+public enum ProcessTerminationStatus : CustomStringConvertible {
+  case exit(Int)
+  case signal(Int)
+
+  public var description: String {
+    switch self {
+    case .exit(let status):
+      return "Exit(\(status))"
+    case .signal(let signal):
+      return "Signal(\(_signalToString(signal)))"
+    }
+  }
+}
+
+
+#if os(Windows)
+public func spawnChild(_ args: [String])
+    -> (process: HANDLE, stdinFD: CInt, stdoutFD: CInt, stderrFD: CInt) {
+  var _stdin: (read: HANDLE?, write: HANDLE?)
+  var _stdout: (read: HANDLE?, write: HANDLE?)
+  var _stderr: (read: HANDLE?, write: HANDLE?)
+
+  var saAttributes: SECURITY_ATTRIBUTES = SECURITY_ATTRIBUTES()
+  saAttributes.nLength = DWORD(MemoryLayout<SECURITY_ATTRIBUTES>.size)
+  saAttributes.bInheritHandle = TRUE
+  saAttributes.lpSecurityDescriptor = nil
+
+  if CreatePipe(&_stdin.read, &_stdin.write, &saAttributes, 0) == FALSE {
+    fatalError("CreatePipe() failed")
+  }
+  if SetHandleInformation(_stdin.write, HANDLE_FLAG_INHERIT, 0) == FALSE {
+    fatalError("SetHandleInformation() failed")
+  }
+
+  if CreatePipe(&_stdout.read, &_stdout.write, &saAttributes, 0) == FALSE {
+    fatalError("CreatePipe() failed")
+  }
+  if SetHandleInformation(_stdout.read, HANDLE_FLAG_INHERIT, 0) == FALSE {
+    fatalError("SetHandleInformation() failed")
+  }
+
+  if CreatePipe(&_stderr.read, &_stderr.write, &saAttributes, 0) == FALSE {
+    fatalError("CreatePipe() failed")
+  }
+  if SetHandleInformation(_stderr.read, HANDLE_FLAG_INHERIT, 0) == FALSE {
+    fatalError("SetHandleInformation() failed")
+  }
+
+  var siStartupInfo: STARTUPINFO = STARTUPINFO()
+  siStartupInfo.cb = DWORD(MemoryLayout<STARTUPINFO>.size)
+  siStartupInfo.hStdError = _stderr.write
+  siStartupInfo.hStdOutput = _stdout.write
+  siStartupInfo.hStdInput = _stdin.read
+  siStartupInfo.dwFlags |= STARTF_USESTDHANDLES
+
+  var piProcessInfo: PROCESS_INFORMATION = PROCESS_INFORMATION()
+
+  if CreateProcessA(args.joined(separator: " "), nil, nil, nil, TRUE,
+                    0, nil, nil, &siStartupInfo, &piProcessInfo) == FALSE {
```

Comment:
Would be nice to see environment and working directory parameters as inputs of the `spawnChild` function; NSTask is going to need a way to propagate that data.

Reply:
That enhancement would have implications for the existing signature, so yes, it would be nice to have those parameters, but, NSTask is going to go through CF and that already DTRT IIRC.